src/libraries/System.Text.Json/tests/System.Text.Json.SourceGeneration.Tests/JsonSerializerContextTests.cs

```diff
@@ -310,6 +310,187 @@ public static IEnumerable<object[]> GetCombiningContextsData()
+                }
+                else
+                {
+                    person = JsonSerializer.Deserialize<Person>(json, options);
+                    VerifySourceGenContextModification(json, person);
+                }
+            }
+        }
+
+        [Fact]
+        public static void SourceGeneratedContextsCanBeModified_JsonTypeInfoOverload()
+        {
+            Person person = new(FirstName: "Jane", LastName: "Doe");
+
+            Test(SourceGenContext.Default.Person);
+            Test(SourceGenContext_Metadata.Default.Person);
+            Test(SourceGenContext_Serialization.Default.Person, isFastPath: true);
+
+            void Test(JsonTypeInfo<Person> typeInfo, bool isFastPath = false)
+            {
+                // This approach of directly modifying src-gen type infos _and_
+                // passing them directly to the serializer has no issues.
+                // Do we need a `IList<JsonTypeInfo> JsonSerializerContext.TypeInfoList` property
+                // on `JsonSerializerContext` or something similar to make it easy to modify all contracts?
+                foreach (JsonPropertyInfo property in typeInfo.Properties)
+                {
+                    Assert.False(isFastPath);
+                    property.Name = property.Name.ToUpperInvariant();
+                }
+
+                string json = JsonSerializer.Serialize(person, typeInfo);
+
+                if (isFastPath)
+                {
+                    JsonTestHelper.AssertJsonEqual(@"{""FirstName"":""Jane"",""LastName"":""Doe""}", json);
+                    Assert.Throws<InvalidOperationException>(() => JsonSerializer.Deserialize(json, typeInfo));
+                }
+                else
+                {
+                    person = JsonSerializer.Deserialize(json, typeInfo);
+                    VerifySourceGenContextModification(json, person);
+                }
+            }
+        }
+
+        [Fact]
+        public static void SourceGeneratedContextsCanBeModified_JsonSerializerContextOverload()
+        {
+            Test(new EditedSourceGenContext_WithCheck());
+            Test(new EditedSourceGenContext_WithCheck_Metadata());
+            Test(new EditedSourceGenContext_WithCheck_Serialization(), isFastPath: true);
+
+            void Test(JsonSerializerContext context, bool isFastPath = false)
+            {
+                Person person = new(FirstName: "Jane", LastName: "Doe");
+                // In this case, user overriden JsonSerializerContext.GetTypeInfo(Type) implementations are consulted.
+                // This is inconsitent with the behavior of `JsonSerializerOptions`-based serializer overloads.
+                // Also see another problem with custom/user JsonSerializerContext.GetTypeInfo(Type) implementations
+                // in `EditTypeInfo(JsonTypeInfo typeInfo, bool checkIfPropsConfigured = false)` below.
+                string json = JsonSerializer.Serialize(person, typeof(Person), context);
```

Comment:
In this case, user overriden JsonSerializerContext.GetTypeInfo(Type) implementations are consulted (given the overload in question). This is inconsitent with the behavior of `JsonSerializerOptions`-based serializer overloads.

Also see another problem with custom/user JsonSerializerContext.GetTypeInfo(Type) implementations in `EditTypeInfo(JsonTypeInfo typeInfo, bool checkIfPropsConfigured = false)` below.

Reply:
Can you elaborate? Unless the user explicitly provides divergent manual implementations for `JsonSerializerContext.GetTypeInfo` and `ITypeInfoResolver.GetTypeInfo` I'm not sure I see the inconsistency. The interface implementation for the base class directly consults virtual method:

https://github.com/dotnet/runtime/blob/bce23abb964d81b6d626f2a2b435549636769a1d/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerContext.cs#L122-L130

Reply:
The `IJsonTypeInfoResolver.GetTypeInfo(Type type, JsonSerializerOptions options)` method isn't called in all serialization scenarios; sometimes the `JsonSerializerContext.GetTypeInfo(Type)` method is called directly. That might be the sole issue here.

The tests are dense but all they do manifest all the issues I'm describing. I'll take a moment to tease out a simplified repro highlighting when user implementations are not being honored.